submissions/shramkoweb/frog-game/app.js

```diff
@@ -0,0 +1,115 @@
+const BLOCK_WIDTH = 101;
+const BLOCK_HEIGHT = 85;
+const ENEMIES_CONF = {
+    MAX_SPEED: 250,
+    MIN_SPEED: 50,
+    OFFSET_Y: 20,
+    SPRITE: 'images/enemy-bug.png',
+    START_X: 0,
+};
+const PLAYER_CONF = {
+    HEIGHT: 60,
+    SPRITE: 'images/char-boy.png',
+    START_X: BLOCK_WIDTH * 2,
+    START_Y: BLOCK_WIDTH * 4,
+    WIDTH: 80,
+};
+
+const Character = function (x, y, sprite) {
+    this.x = x;
+    this.y = y;
+    this.sprite = sprite;
+};
+
+Character.prototype.render = function () {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+const Enemy = function (x, y, player) {
+    Character.call(this, x, y, ENEMIES_CONF.SPRITE);
+    this.speed = this.getRandomSpeed();
+    this.player = player;
+};
+
+Enemy.prototype = Object.create(Character.prototype);
+
+Enemy.prototype.getRandomSpeed = function () {
+    return Math.random() * (ENEMIES_CONF.MAX_SPEED - ENEMIES_CONF.MIN_SPEED) + ENEMIES_CONF.MIN_SPEED;
+};
+
+Enemy.prototype.isCollision = function () {
+    if (
+        this.player.x < this.x + PLAYER_CONF.WIDTH &&
+        this.player.x + PLAYER_CONF.WIDTH > this.x &&
+        this.player.y < this.y + PLAYER_CONF.HEIGHT &&
+        this.player.y + PLAYER_CONF.HEIGHT > this.y
+    ) {
+        alert('You lose!')
+        this.player.x = PLAYER_CONF.START_X;
+        this.player.y = PLAYER_CONF.START_Y;
+    }
+};
+
+Enemy.prototype.update = function (dt) {
```

Comment:
The update method should only do rendering. Better to move the collision check from this method

[Separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns)


You have commented part of code in `engine.js` file 

```
function update(dt) {
        updateEntities(dt);
        // checkCollisions();
    }
```
So it better to implement colliisions check outside of update, make a function checkCollisions, where you will call a method with collision checking for all entities

Reply:
I need create one more method and call my isCollision inside them?

Reply:
This is comment from `engine.js`

```
/* This function is called by main (our game loop) and itself calls all
     * of the functions which may need to update entity's data. Based on how
     * you implement your collision detection (when two entities occupy the
     * same space, for instance when your character should die), you may find
     * the need to add an additional function call here. For now, we've left
     * it commented out - you may or may not want to implement this
     * functionality this way (you could just implement collision detection
     * on the entities themselves within your app.js file).
     */
```

Main point of this is to separete different logic in two separate methods, and than implement checkCollisions() in `engine.js` similary as updateEntities().


So update method do only coordinate reset,increasing speed(for Enemy).

If you dont understand me, i am open to your questions)

Reply:
thx